### PR TITLE
Add tests for MicrosoftGraph extension

### PR DESCRIPTION
**Pull request by ConnorMcMahon**
I have already added tests for all but the most complex of the Excel bindings. This structure of splitting out the Graph access into a *Client class and the business logic in a *Manager class will be used for the other bindings as well. This gives us isolation of the calls to Microsoft Graph, a clear surface area definition of what we are trying to do in graph, and a mockable interface.

NOTE: This approach does mean that in order to get true test coverage, we would want either end-to-end tests that will require the user to give valid credentials, or extensive tests of each of the *Client classes, which would likely also need credentials. EndToEnd tests would probably be preferred.

### CODE REVIEW

File: WebJobs.Extensions.O365/Bindings/ExcelAsyncCollector.cs [hunk 1]
```diff
@@ -55,35 +58,66 @@ public Task AddAsync(JObject item, CancellationToken cancellationToken = default
...
-                    await this.client.AddRow(this.attribute, row);
+                    await _manager.AddRow(this._attribute, row);
                 }
             }
 
-            this.rows.Clear();
+            this._rows.Clear();
+        }
+
+        public JObject GetConsolidatedRows(Collection<JObject> rows)
```
Comment (ConnorMcMahon) [inline WebJobs.Extensions.O365/Bindings/ExcelAsyncCollector.cs]: This function and its corresponding logic in FlushAsync was necessary to give Collector<POCO> parity with POCO[] and List<POCO>. It also should have the advantage of being more performant than updating a single row at a time.

File: test/MicrosoftGraph.Tests/ExcelTests.cs [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+    using Microsoft.Azure.WebJobs.Extensions.Token.Tests;
+    using Microsoft.Azure.WebJobs.Extensions.MicrosoftGraph.Services;
+    using Moq;
+    using Microsoft.Graph;
+    using System.Collections.Generic;
+    using Newtonsoft.Json.Linq;
+    using System.Linq;
+    using Newtonsoft.Json;
+
+    public class ExcelTestsEndToEnd
```
Comment (ConnorMcMahon) [inline test/MicrosoftGraph.Tests/ExcelTests.cs]: My general approach for input bindings is to define a function that has an attribute on a parameter input. When the jobhost runs the function, the binding should resolve into it's final value, allowing me to set it to some global variable that the test can check to make sure that it is correct.

For output bindings, the main concern isn't the binding's end value, as much as the fact that the GraphServiceClient is hit with the correct values. This means that mock.Verify on the Post and Patch methods on Excel client are what actually test the validity of the output binding.

File: test/MicrosoftGraph.Tests/OneDriveTests.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    using System.IO;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.WebJobs.Extensions.MicrosoftGraph.Services;
+    using Microsoft.Azure.WebJobs.Extensions.Token.Tests;
+    using Microsoft.Graph;
+    using Moq;
+    using Xunit;
+
+    public class OneDriveTests
```
Comment (ConnorMcMahon) [inline test/MicrosoftGraph.Tests/OneDriveTests.cs]: I didn't do many tests here as after talking with @MikeStall, it became clear that there will need to be changes to these bindings.

File: WebJobs.Extensions.O365/Services/ExcelClient.cs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    using Newtonsoft.Json.Linq;
+
+    /// <summary>
+    /// Helper for calling onto Excel (MS) Graph
+    /// </summary>
+    internal class ExcelClient : IExcelClient
+    {
+        private Task<IGraphServiceClient> _client;
+
+        public ExcelClient(Task<IGraphServiceClient> client)
```
Comment (MikeStall) [inline WebJobs.Extensions.O365/Services/ExcelClient.cs]: >Task<IGraphServiceClient> [](start = 27, length = 25)

Constructors should  not take a Task<> parameter. The caller should await it. And pass in the result.

File: WebJobs.Extensions.O365/Services/IExcelClient.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.Azure.WebJobs.Extensions.MicrosoftGraph.Tests")]
+namespace Microsoft.Azure.WebJobs.Extensions.MicrosoftGraph.Services
+{
+    using System.Threading.Tasks;
+    using Microsoft.Graph;
+    using Newtonsoft.Json.Linq;
+
+    interface IExcelClient
```
Comment (MikeStall) [inline WebJobs.Extensions.O365/Services/IExcelClient.cs]: Does this become a wrapper around the Graph SDK's Excel client?  Can we avoid this (is their client mockable)

Comment (ConnorMcMahon) [reply WebJobs.Extensions.O365/Services/IExcelClient.cs]: It is a wrapper. Their client, while it does have an interface, would be fairly painful to mock, as it creates deep call chains using the builder pattern. 

I also think having the wrapper makes the code cleaner. If you look at the implementation in ExcelClient (or any of the other client classes), it is all Graph SDK stuff that is relatively complicated and doesn't really match the level of abstraction as the business logic done by the rest of the code. Also, many of those graph calls are made several times in the code, so putting all of this logic in these functions allowed for a lot more reusability.

File: WebJobs.Extensions.O365/Services/ExcelClient.cs (hunk 4 above)
Comment (ConnorMcMahon) [reply WebJobs.Extensions.O365/Services/ExcelClient.cs]: Both feel kind of dirty to me, but its good to know this is the lesser of two evils :)

File: WebJobs.Extensions.O365/Config/MicrosoftGraphExtensionConfig.cs [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+
+            public Converters(ServiceManager parent, GraphWebhookConfig webhookConfig)
+            {
+                _serviceManager = parent;
+                _webhookConfig = webhookConfig;
+            }
+
+            public IAsyncCollector<JObject> CreateCollector(ExcelAttribute attr)
+            {
+                var service = _serviceManager.GetExcelService(attr).Result;
```
Comment (ConnorMcMahon) [inline WebJobs.Extensions.O365/Config/MicrosoftGraphExtensionConfig.cs]: I'm not sure I like the forced synchronization here, as the collectors don't really need the service until they are flushed at the end of the function call anyways. However, without allowing a task as an input to the constructor, there is no good way to force this. @MikeStall, thoughts here?

Comment (MikeStall) [reply WebJobs.Extensions.O365/Config/MicrosoftGraphExtensionConfig.cs]: Don't call .Result - that can deadlock. 
Queue has this same problem. See [https://github.com/Azure/azure-webjobs-sdk/blob/e915ee574136207cab3b01cac7139bf09b013ace/src/Microsoft.Azure.WebJobs.Host/Queues/Bindings/QueueBindingProvider.cs#L190](https://github.com/Azure/azure-webjobs-sdk/blob/e915ee574136207cab3b01cac7139bf09b013ace/src/Microsoft.Azure.WebJobs.Host/Queues/Bindings/QueueBindingProvider.cs#L190) 
Actually, we should add a method on BindToCollector to make Async easier here. Can you file a bug for that? 




---
In reply to: [141749488](https://github.com/Azure/azure-functions-microsoftgraph-extension/pull/13#discussion_r141749488) [](ancestors = 141749488)

**Pull request merged**